src/System.Web.OData/OData/Query/Expressions/SelectExpandBinder.cs

```diff
@@ -187,30 +187,64 @@ internal Expression CreatePropertyValueExpression(IEdmEntityType elementType, IE
-                            nullablePropertyValue)
-                        : nullablePropertyValue;
-                // TODO: Implement proper support for $select/$expand after $apply
-                Expression filterPredicate = FilterBinder.Bind(null, filterClause, clrElementType, _context.RequestContainer);
-                MethodCallExpression filterResult = Expression.Call(
-                    ExpressionHelperMethods.QueryableWhereGeneric.MakeGenericMethod(clrElementType),
-                    filterSource,
-                    filterPredicate);
+                Expression filterResult = nullablePropertyValue;
+
+                if (collection)
+                {
+                    Expression filterSource =
+                        typeof(IEnumerable).IsAssignableFrom(source.Type.GetProperty(propertyName).PropertyType)
+                            ? Expression.Call(
+                                ExpressionHelperMethods.QueryableAsQueryable.MakeGenericMethod(clrElementType),
+                                nullablePropertyValue)
+                            : nullablePropertyValue;
+                    
+                    // TODO: Implement proper support for $select/$expand after $apply
+                    Expression filterPredicate = FilterBinder.Bind(null, filterClause, clrElementType, _context.RequestContainer);
+                    filterResult = Expression.Call(
+                        ExpressionHelperMethods.QueryableWhereGeneric.MakeGenericMethod(clrElementType),
+                        filterSource,
+                        filterPredicate);
+
+                    nullablePropertyType = filterResult.Type;
+                }
+                else if(_settings.HandleReferenceNavigationPropertyExpandFilter)
+                {                    
+                    var filterLambdaExpression = FilterBinder.Bind(null, filterClause, clrElementType, _context.RequestContainer) as LambdaExpression;
+                    if(filterLambdaExpression == null)
+                    {
+                        throw new ODataException(Error.Format(SRResources.ExpandFilterExpressionNotLambdaExpression,
+                            property.Name, nameof(LambdaExpression)));
+                    }
+
+                    var filterParameter = filterLambdaExpression.Parameters.FirstOrDefault();
+                    if(filterParameter == null)
+                    {   
+                        //Never seen this, but just to be safe...
+                        throw new ODataException(Error.Format(SRResources.ExpandFilterExpressionLambdaExpressionNoParameter,
+                            property.Name));                            
+                    }
+
+                    var predicateExpression = new ReferenceNavigationPropertyExpandFilterVisitor(filterParameter, nullablePropertyValue).Visit(filterLambdaExpression.Body);
+
+                    // predicateExpression == true ? nullablePropertyValue : null
```

Comment:
what's this line for?

Reply:
describes what kind of expression is created. will update comment